File: attractors/2D/Go/Quadruptwo.go
/*
 *  (C) Copyright 2025  Pavel Tisnovsky
 *
 *  All rights reserved. This program and the accompanying materials
 *  are made available under the terms of the Eclipse Public License v1.0
 *  which accompanies this distribution, and is available at
 *  http://www.eclipse.org/legal/epl-v10.html
 *
 *  Contributors:
 *      Pavel Tisnovsky
 */

package main

import (
	"attractors/bitmap"
	"image"
	"math"
)

const ImageWidth = 800
const ImageHeight = 600
const Maxiter = 1000000

func sqr(x float64) float64 {
	return x * x
}

func sign(x float64) float64 {
	if x > 0 {
		return 1
	}
	return -1
}

func quadruptwo(x, y, a, b, c float64) (float64, float64) {
	xn := y - sign(x)*math.Sin(math.Log(math.Abs(b*x-c)))*math.Atan(sqr(math.Log(math.Abs(c*x-b))))
	yn := a - x
	return xn, yn
}

func calcSystem(
	width, height, maxiter int,
	A, B, C, scale, xOffset, yOffset, maxFactor float64) *image.RGBA {

	img := image.NewRGBA(image.Rect(0, 0, width, height))

	bmp := bitmap.CreateBitmap(width, height)
	maxPoints := maxiter
	settleDownPoints := 100

	x := 0.0
	y := 0.0

	for i := range maxPoints {
		xn, yn := quadruptwo(x, y, A, B, C)
		xi := int(float64(width)/2 + scale*xn + xOffset)
		yi := int(float64(height)/2 + scale*yn + yOffset)
		if i > settleDownPoints {
			if xi >= 0 && yi >= 0 && xi < int(width) && yi < int(height) {
				bmp[yi][xi] += 1.0
			}
		}
		x, y = xn, yn
	}

	bitmap.DrawBitmapToImage(bmp, img, maxFactor)
	return img
}

func main() {
	{
		const A = 3.1
		const B = 1.8
		const C = -0.9
		img := calcSystem(ImageWidth, ImageHeight, Maxiter, A, B, C, 9.0, 0, 0, 1/3.0)
		bitmap.SaveImage(img, "quadruptwo.png")
	}
}

File: attractors/2D/Pygame/Quadruptwo.py
"""Výpočet a vykreslení podivného atraktoru nazvaného Quadruptwo."""

#
#  (C) Copyright 2024  Pavel Tisnovsky
#
#  All rights reserved. This program and the accompanying materials
#  are made available under the terms of the Eclipse Public License v1.0
#  which accompanies this distribution, and is available at
#  http://www.eclipse.org/legal/epl-v10.html
#
#  Contributors:
#      Pavel Tisnovsky
#

from math import atan, copysign, sin
from math import log as ln
from typing import Tuple

from renderer import render_attractor
from ui import event_loop, initialize

# window settings
WINDOW_WIDTH = 800
WINDOW_HEIGHT = 600
WINDOW_TITLE = "Quadruptwo dynamical system"


def sqr(x: float) -> float:
    """Square calculation."""
    return x * x


def quadruptwo(
    x: float, y: float, A: float, B: float, C: float, D: float
) -> Tuple[float, float]:
    """Calculate next point in the strange attractor."""
    x_dot = y - copysign(1, x) * sin(ln(abs(B * x - C))) * atan(sqr(ln(abs(C * x - B))))
    y_dot = A - x
    return x_dot, y_dot


def main() -> None:
    # initialize user interface based on Pygame
    display, surface, clock = initialize(WINDOW_TITLE, WINDOW_WIDTH, WINDOW_HEIGHT)

    # redraw the whole dynamical system
    render_attractor(
        surface=surface,
        x=0,
        y=0,
        scale=9.0,
        x_offset=0,
        y_offset=0,
        max_points=1000000,
        settle_down_points=10,
        attractor_formula=quadruptwo,
        contrast=1 / 7.0,
        filename="quadruptwo.png",
        A=3.1,
        B=1.8,
        C=-0.9,
        D=0.2277,
    )

    # and enter the event loop
    event_loop(display, surface, clock)


if __name__ == "__main__":
    main()
